tests/ci/check_so_interface.py

```diff
@@ -10,42 +10,210 @@
+        found_args = [arg.arg for arg in func_node.args.args]
+        found_defaults = [ast.unparse(d) if d else None for d in func_node.args.defaults]
+        
+        # Extract arguments from the spec function
+        spec_sig = ast.parse(f"def {func_name}{ast.get_source_segment(spec_func, spec_func.args)}:pass").body[0]
+        spec_args_list = [arg.arg for arg in spec_sig.args.args]
+        spec_defaults = [ast.unparse(d) if d else None for d in spec_sig.args.defaults]
+        
+        # Compare arguments
+        if found_args != spec_args_list:
+            print(f"Mismatch in arguments for function {func_name}")
+            print(f"Found: {found_args}")
+            print(f"Expected: {spec_args_list}")
+            return False
+        
+        # Compare default values
+        if found_defaults != spec_defaults:
+            print(f"Mismatch in default values for function {func_name}")
+            print(f"Found: {found_defaults}")
+            print(f"Expected: {spec_defaults}")
             return False
-        # Check argument type annotation if specified
-        if 'type' in expected: #if there is a type annotation
-            if actual.annotation is None: #check if annotation of actual -> false
-                return False
-            if not isinstance(actual.annotation, ast_args.Name) or actual.annotation.id != expected['type']: #if the annotation of actual doesn't match expe
-                return False
-        # Check default value if specified
-        if 'default' in expected:
-            if actual.default is None:
-                return False
-            if not isinstance(actual.default, ast.Constant) or actual.default.value != expected['default']:
-                return False
-    # Check for *args
-    # Check for **kwargs
+    
     return True
+
 if __name__ == "__main__":   #if this is the main thing being run
     for fname, spec in interface.items():   #loop over file names "fname" as keys and "spec" as values in dictionary called "interface"
+        print(interface.items())
         with open(fname, 'r') as fh:   #opens file "fname", in read mode 'r', to be used in code as "fh"
             tree = ast.parse(fh.read()) #parsing contents of file into abstract syntax tree
-        for key in spec:
-            # check this function/class is in the code
-            assert key in tree
-            # compare its arguments
-            if isinstance(tree[key], ast.FunctionDef):
-                assert _compare_args(spec[key], tree[key])
-            elif isinstance(tree[key], ast.ClassDef):
-                # could make this recursive maybe
-                for meth in spec[key]: #iterating over methods specified in spec[key]
-                    assert meth in tree[key]
-                    assert _compare_args(spec[key][meth], tree[key][meth])
+            notfound = []
+            found = {}
+            dump = ast.dump(tree)
+        for node in ast.walk(tree):
```

Comment:
the first layer in the tree is usually going to be made up of `ClassDef` objects (though not always), which in turn contain `FunctionDef`'s. The original sketch had a check for this.